ResourceCollection NPE in list()

**Jetty version**
9.4.x

**Description**
```
Caused by: java.lang.NullPointerException: Cannot read the array length because "<local3>" is null
    at java.util.Collections.addAll (Collections.java:5592)
    at org.eclipse.jetty.util.resource.ResourceCollection.list (ResourceCollection.java:423)
    at org.eclipse.jetty.util.resource.Resource.getAllResources (Resource.java:886)
    at org.eclipse.jetty.annotations.AnnotationParser.parseDir (AnnotationParser.java:777)
    at org.eclipse.jetty.annotations.AnnotationParser.parse (AnnotationParser.java:735)
    at org.eclipse.jetty.annotations.AnnotationConfiguration$ParserTask.call (AnnotationConfiguration.java:172)
    at org.eclipse.jetty.annotations.AnnotationConfiguration$1.run (AnnotationConfiguration.java:465)
```

Looks like the last commit broken jetty-maven-plugin's classloader.

It failed, consistently, across all JDKs.

```
Exception in thread "Thread-190" java.lang.NoClassDefFoundError: org/testcontainers/utility/PathUtils
	at org.testcontainers.utility.MountableFile.lambda$deleteOnExit$0(MountableFile.java:292)
       at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: org.testcontainers.utility.PathUtils
	at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
```

--- a/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java
+++ b/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java
@@ -571,11 +571,7 @@ public class PathResource extends Resource
             int size = entries.size();
             return entries.toArray(new String[size]);
         }
-        catch (DirectoryIteratorException e)
-        {
-            LOG.debug(e);
-        }
-        catch (IOException e)
+        catch (DirectoryIteratorException | IOException e)
         {
             LOG.debug(e);
         }

--- a/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
+++ b/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
@@ -440,7 +440,6 @@ public class ResourceCollection extends Resource
             String[] list = r.list();
             if (list != null)
                 Collections.addAll(set, list);
-            Collections.addAll(set, r.list());
         }
         String[] result = set.toArray(new String[0]);
         Arrays.sort(result);

--- a/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
+++ b/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
@@ -256,51 +256,40 @@ public class ResourceCollection extends Resource
             return this;
         }
 
-        Resource resource = null;
         ArrayList<Resource> resources = null;
-        int i = 0;
-        for (; i < _resources.length; i++)
+
+        // Attempt a simple (single) Resource lookup that exists
+        for (Resource res : _resources)
         {
-            resource = _resources[i].addPath(path);
-            if (resource.exists())
+            Resource r = res.addPath(path);
+            if (!r.exists())
+                continue;
+
+            if (!r.isDirectory())
+                return r;
+
+            if (resources == null)
+                resources = new ArrayList<>();
+            resources.add(r);
+        }
+
+        if (resources == null)
+        {
+            return null; /* TODO this is not allowed in 10.  Instead do:
+            for (Resource res : _r1esources)
             {
-                if (resource.isDirectory())
-                {
-                    break;
-                }
-                return resource;
+                Resource r = res.addPath(path);
+                if (r.exists())
+                    return r;
             }
+            return EmptyResource.INSTANCE;
+            */
         }
 
-        for (i++; i < _resources.length; i++)
-        {
-            Resource r = _resources[i].addPath(path);
-            if (r.exists() && r.isDirectory())
-            {
-                if (resources == null)
-                {
-                    resources = new ArrayList<>();
-                }
+        if (resources.size() == 1)
+            return resources.get(0);
 
-                if (resource != null)
-                {
-                    resources.add(resource);
-                    resource = null;
-                }
-
-                resources.add(r);
-            }
-        }
-
-        if (resource != null)
-        {
-            return resource;
-        }
-        if (resources != null)
-        {
-            return new ResourceCollection(resources.toArray(new Resource[0]));
-        }
-        return null;
+        return new ResourceCollection(resources.toArray(new Resource[0]));
     }
 
     @Override

--- a/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
+++ b/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
@@ -434,10 +434,12 @@ public class ResourceCollection extends Resource
     public String[] list()
     {
         assertResourcesSet();
-
         HashSet<String> set = new HashSet<>();
         for (Resource r : _resources)
         {
+            String[] list = r.list();
+            if (list != null)
+                Collections.addAll(set, list);
             Collections.addAll(set, r.list());
         }
         String[] result = set.toArray(new String[0]);

--- a/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
+++ b/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
@@ -256,40 +256,51 @@ public class ResourceCollection extends Resource
             return this;
         }
 
+        Resource resource = null;
         ArrayList<Resource> resources = null;
-
-        // Attempt a simple (single) Resource lookup that exists
-        for (Resource res : _resources)
+        int i = 0;
+        for (; i < _resources.length; i++)
         {
-            Resource r = res.addPath(path);
-            if (!r.exists())
-                continue;
-
-            if (!r.isDirectory())
-                return r;
-
-            if (resources == null)
-                resources = new ArrayList<>();
-            resources.add(r);
-        }
-
-        if (resources == null)
-        {
-            return null; /* TODO this is not allowed in 10.  Instead do:
-            for (Resource res : _r1esources)
+            resource = _resources[i].addPath(path);
+            if (resource.exists())
             {
-                Resource r = res.addPath(path);
-                if (r.exists())
-                    return r;
+                if (resource.isDirectory())
+                {
+                    break;
+                }
+                return resource;
             }
-            return EmptyResource.INSTANCE;
-            */
         }
 
-        if (resources.size() == 1)
-            return resources.get(0);
+        for (i++; i < _resources.length; i++)
+        {
+            Resource r = _resources[i].addPath(path);
+            if (r.exists() && r.isDirectory())
+            {
+                if (resources == null)
+                {
+                    resources = new ArrayList<>();
+                }
 
-        return new ResourceCollection(resources.toArray(new Resource[0]));
+                if (resource != null)
+                {
+                    resources.add(resource);
+                    resource = null;
+                }
+
+                resources.add(r);
+            }
+        }
+
+        if (resource != null)
+        {
+            return resource;
+        }
+        if (resources != null)
+        {
+            return new ResourceCollection(resources.toArray(new Resource[0]));
+        }
+        return null;
     }
 
     @Override

--- a/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceCollectionTest.java
+++ b/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceCollectionTest.java
@@ -22,6 +22,7 @@ import java.io.BufferedReader;
 import java.io.File;
 import java.io.InputStreamReader;
 import java.nio.file.Path;
+import java.util.Arrays;
 
 import org.eclipse.jetty.toolchain.test.FS;
 import org.eclipse.jetty.toolchain.test.MavenTestingUtils;
@@ -32,7 +33,9 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -172,6 +175,19 @@ public class ResourceCollectionTest
             Path destPath = workdir.getPathFile("bar");
             coll.copyTo(destPath.toFile());
         });
+    }
+
+    @Test
+    public void testList() throws Exception
+    {
+        ResourceCollection rc1 = new ResourceCollection(
+            Resource.newResource("src/test/resources/org/eclipse/jetty/util/resource/one/"),
+            Resource.newResource("src/test/resources/org/eclipse/jetty/util/resource/two/"),
+            Resource.newResource("src/test/resources/org/eclipse/jetty/util/resource/three/"));
+
+        assertThat(Arrays.asList(rc1.list()), contains("1.txt", "2.txt", "3.txt", "dir/"));
+        assertThat(Arrays.asList(rc1.addPath("dir").list()), contains("1.txt", "2.txt", "3.txt"));
+        assertThat(rc1.addPath("unknown").list(), nullValue());
     }
 
     @Test

--- a/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceCollectionTest.java
+++ b/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceCollectionTest.java
@@ -187,7 +187,7 @@ public class ResourceCollectionTest
 
         assertThat(Arrays.asList(rc1.list()), contains("1.txt", "2.txt", "3.txt", "dir/"));
         assertThat(Arrays.asList(rc1.addPath("dir").list()), contains("1.txt", "2.txt", "3.txt"));
-        assertThat(rc1.addPath("unknown"), nullValue());
+        assertThat(rc1.addPath("unknown").list(), nullValue());
         // TODO for jetty-10 assertThat(rc1.addPath("unknown").list(), nullValue());
     }
 

--- a/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceCollectionTest.java
+++ b/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceCollectionTest.java
@@ -187,7 +187,8 @@ public class ResourceCollectionTest
 
         assertThat(Arrays.asList(rc1.list()), contains("1.txt", "2.txt", "3.txt", "dir/"));
         assertThat(Arrays.asList(rc1.addPath("dir").list()), contains("1.txt", "2.txt", "3.txt"));
-        assertThat(rc1.addPath("unknown").list(), nullValue());
+        assertThat(rc1.addPath("unknown"), nullValue());
+        // TODO for jetty-10 assertThat(rc1.addPath("unknown").list(), nullValue());
     }
 
     @Test

